Add documentation about changing cluster node hostname (#1180)
scality/metalk8s: docs/operation/changing_node_hostname.rst

Changing the hostname of a MetalK8s node
========================================

#. On the node, change the hostname:

   .. code-block:: shell

      $ hostnamectl set-hostname <New hostname>
      $ systemctl restart systemd-hostnamed

#. Check that the change is taken into account.

   .. code-block:: shell

      $ hostnamectl status

>       Static hostname: controlplane2

Reviewer: Make sure to use the same placeholders across the document, to help users see what should be verified.

```suggestion
      Static hostname: <new_hostname>
```

>       Pretty hostname: ControlPlane2

Reviewer: Where does this come from? Is it automatically set? Did you choose it when running `set-hostname`?

Author: yes, I replaced it by <New hostname> as requested in the previous comment

        Icon name: computer-vm
         Chassis: vm
      Machine ID: 5003025f93c1a84914ea5ae66519c100
         Boot ID: f28d5c64f06c48a3a775e24c4f03d00c
         Virtualization: kvm
      Oerating System: CentOS Linux 7 (Core)
         CPE OS Name: cpe:/o:centos:centos:7
             Kernel: Linux 3.10.0-957.12.2.el7.x86_64
         Architecture: x86-64

> #. On the bootstrap node, check the hostname edition allowed a change of status

Reviewer: ```suggestion
#. On the bootstrap node, check the hostname edition incurred a change of status.
```

>    on the bootstrap. The node must be in a **NotReady** status.

Reviewer: ```suggestion
   The edited node must be in a **NotReady** state.
```

>    status on the bootstrap. The node must be in a **NotReady** status.

Reviewer: ```suggestion
   status. The edited node must be in a **NotReady** status.
```

   .. code-block:: shell

      $ kubectl get nodes

>       node2    NotReady   etcd,master                   19h       v1.11.7

Reviewer: You'll probably have more than just `node2` when running `kubectl get nodes`. Either add the _usual_ values in the output, or change the command to `kubectl get node <node_name>`.

> #. Change the name of the node in the ``yaml`` file used to create it.

Reviewer: Maybe add a command to generate a valid YAML manifest from the exisiting `Node` object... Will check how to do that.

Reviewer: Also add a reference to the Quickstart section that details Node manifests.

>    .. code-block:: yaml

Reviewer: ditto, `parsed-literal` is needed here

      apiVersion: v1
      kind: Node
      metadata:
        name: <New_node_name>
        annotations:
          metalk8s.scality.com/ssh-key-path: /etc/metalk8s/pki/salt-bootstrap
          metalk8s.scality.com/ssh-host: <node control-plane IP>
          metalk8s.scality.com/ssh-sudo: 'false'
        labels:
          metalk8s.scality.com/version: '|release|'
          <role labels>
      spec:
        taints: <taints>

   Then apply the configuration:

   .. code-block:: shell

      $ kubectl apply -f new-node-creation.yaml

#. Delete the old node (here ``node2``):

   .. code-block:: shell

      $ kubectl delete node node2

> #. Log on the salt master:

Reviewer: ```suggestion
#. Open a terminal into the :term:`Salt master` container:
```

   .. code-block:: shell

>       $ kubectl -it -n kube-system exec salt-master-bootstrap bash

Reviewer: Be careful, static Pod names depend on the actual Node name where they exist.

```suggestion
      $ kubectl -it exec salt-master-<bootstrap_node_name> -n kube-system -c salt-master bash
```

> #. Delete the old node key:

Reviewer: ```suggestion
#. Delete the now obsolete :term:`Salt minion` key for the changed Node:
```

   .. code-block:: shell

      $ salt-key -d <node_key_name>

> #. Run the new node name deployment:

Reviewer: ```suggestion
#. Re-run the deployment for the edited Node:
```

>    .. code-block:: shell

Reviewer: You need a `.. parsed-literal::` directive to have the substition on `|release|`

     $ salt-run state.orchestrate metalk8s.orchestrate.deploy_node \
       saltenv=metalk8s-|release| \
       pillar='{"orchestrate": {"node_name": "<new-node-name>"}}'

          Summary for bootstrap_master
          -------------
          Succeeded: 11 (changed=9)
          Failed:     0
          -------------
          Total states run:     11
          Total run time:  132.435 s

> #. On the bootstrap node, restart the kubelet service:

Reviewer: Probably not on Bootstrap:

```suggestion
#. On the edited node, restart the :term:`kubelet` service:
```

Changing the hostname of a MetalK8s node
========================================

#. On the node, change the hostname:

   .. code-block:: shell

      $ hostnamectl set-hostname <New hostname>
      $ systemctl restart systemd-hostnamed

#. Check that the change is taken into account.

   .. code-block:: shell

      $ hostnamectl status

      Static hostname: <New hostname>
      Pretty hostname: <New hostname>
        Icon name: computer-vm
         Chassis: vm
      Machine ID: 5003025f93c1a84914ea5ae66519c100
         Boot ID: f28d5c64f06c48a3a775e24c4f03d00c
         Virtualization: kvm
      Oerating System: CentOS Linux 7 (Core)
         CPE OS Name: cpe:/o:centos:centos:7
             Kernel: Linux 3.10.0-957.12.2.el7.x86_64
         Architecture: x86-64

#. On the bootstrap node, check the hostname edition incurred a change of
   status on the bootstrap. The edited node must be in a **NotReady** status.

   .. code-block:: shell

      $ kubectl get <node_name>

>       node2    NotReady   etcd,master                   19h       v1.11.7

Reviewer: ```suggestion
      <node_name>   NotReady   etcd,master                   19h       v1.11.7
```

#. Change the name of the node in the ``yaml`` file used to create it.
   Refer to :ref:`Creating a manifest` for more information.

   .. parsed-literal::

      apiVersion: v1
      kind: Node
      metadata:
        name: <New_node_name>
        annotations:
          metalk8s.scality.com/ssh-key-path: /etc/metalk8s/pki/salt-bootstrap
          metalk8s.scality.com/ssh-host: <node control-plane IP>
          metalk8s.scality.com/ssh-sudo: 'false'
        labels:
          metalk8s.scality.com/version: '|release|'
          <role labels>
      spec:
        taints: <taints>

   Then apply the configuration:

   .. code-block:: shell

>       $ kubectl apply -f new-node-creation.yaml

Reviewer: `new-node-creation.yaml`:

- not a good name (the YAML is a manifest, it describes a Node, not its creation)
- not mentioned before, so could be replaced with `<path to edited manifest>`

> #. Delete the old node (here ``node2``):

Reviewer: ```suggestion
#. Delete the old node (here ``<node_name>``):
```

   .. code-block:: shell

>       $ kubectl delete node node2

Reviewer: ```suggestion
      $ kubectl delete node <node_name>
```

#. Open a terminal into the :term:`Salt master` container:

   .. code-block:: shell

      $ kubectl -it exec salt-master-<bootstrap_node_name> -n kube-system -c salt-master bash

#. Delete the now obsolete :term:`Salt minion` key for the changed Node:

   .. code-block:: shell

>       $ salt-key -d <node_key_name>

Reviewer: ```suggestion
      $ salt-key -d <node_name>
```
